### PR TITLE
Only show relevant facilities in some interface drop-down menus

### LINKED ISSUE #28039: Powder Diffraction Reduction change instrument has invalid options
### Expected behavior
When oppened for the first time or when choosing to change the instrument on the Powder Diffraction Reduction GUI a dialog box opens with several options. these options should all be valid
### Actual behavior
In the dialog box for facility you can select `MLZ`, `ISIS`, or `HFIR`. none of these have any instruments associated and if chosen will just open up the dialog again and ask untill you select a valid option.
Also selecting cancel will close the whole gui instead of closing the dialog box like one would expect.
### Steps to reproduce the behavior
1. open workbnch
2. `Interfaces->Diffraction->Powder Diffraction Reduction`
3. If this is your first time opening the gui the dialog box should open, if not go to `Tools->Change Instrument`
4. Select `SNS` and `NOM`. the gui should open normaly.
5. `Tools->Change Instrument`
6. Select `MLZ`, there is no instrument option
7. Click OK, the gui goes blank and the dialog open again.
8. Try with `ISIS` or `HFIR`, the same result should happen.
9. Go back to `SNS` and `NOM`.
10. Open the change instrument dialog.
11. Click cancel, instead of closing the dialog it closes the gui.
### Platforms affected
Windows

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
One line bug fix to show only the facilities which house the instruments for which the interface was designed. This affects the following interfaces:
- Diffraction > Powder Diffraction Reduction
- SANS > ORNL_SANS
- Direct > DGS Reduction

To Test
1. open workbench
2. Open each of the interfaces above (e.g.  Interfaces->Diffraction->Powder Diffraction Reduction)
If this is your first time opening the gui the dialog box should open, if not go to Tools->Change Instrument
3. Change the facilities in the drop-down menu - there should be no facilities with an empty instrument list
 
Fixes #28039

### CODE REVIEW

**Comment by Harrietbrown** [comment]
Powder Diffraction Reduction and DGS Reduction are fine, but ORNL SANS still has HFIR which results in this error if chosen:

```
Traceback (most recent call last):
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\applications\workbench\workbench\app\mainwindow.py", line 422, in <lambda>
    action.triggered.connect(lambda checked_py, script=script: self.launch_custom_python_gui(script))
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\applications\workbench\workbench\app\mainwindow.py", line 381, in launch_custom_python_gui
    self.interface_executor.execute(open(filename).read(), filename)
  File "c:\users\wey38795\documents\mantid_builds\master\raw\mantid\qt\python\mantidqt\widgets\codeeditor\execution.py", line 143, in execute
    executor.execute(code_str, filename, flags, self.globals_ns)
  File "C:/Users/wey38795/Documents/mantid_builds/Master/raw/mantid/scripts/ORNL_SANS.py", line 15, in <module>
    if reducer.setup_layout(load_last=True):
  File "C:\Users\wey38795\Documents\mantid_builds\Master\raw\mantid\scripts\Interface\reduction_application.py", line 164, in setup_layout
    self._interface = instrument_factory(self._instrument, settings=self.general_settings)
  File "C:\Users\wey38795\Documents\mantid_builds\Master\raw\mantid\scripts\Interface\reduction_gui\instruments\instrument_factory.py", line 40, in instrument_factory
    return INSTRUMENT_DICT[facility][instrument](instrument, settings=settings)
  File "C:\Users\wey38795\Documents\mantid_builds\Master\raw\mantid\scripts\Interface\reduction_gui\instruments\hfir_interface_dev.py", line 58, in __init__
    self.attach(SampleDataWidget(settings = self._settings, data_proxy=DataProxy))
  File "C:\Users\wey38795\Documents\mantid_builds\Master\raw\mantid\scripts\Interface\reduction_gui\widgets\sans\hfir_sample_data.py", line 240, in __init__
    self.set_state(m)
  File "C:\Users\wey38795\Documents\mantid_builds\Master\raw\mantid\scripts\Interface\reduction_gui\widgets\sans\hfir_sample_data.py", line 283, in set_state
    self._direct_beam(state=state.calculation_method)
  File "C:\Users\wey38795\Documents\mantid_builds\Master\raw\mantid\scripts\Interface\reduction_gui\widgets\sans\hfir_sample_data.py", line 365, in _direct_beam
    self._replace_method(DirectBeam(self, state=state, settings=self._settings,
  File "C:\Users\wey38795\Documents\mantid_builds\Master\raw\mantid\scripts\Interface\reduction_gui\widgets\sans\hfir_sample_data.py", line 42, in __init__
    self.initialize_content()
  File "C:\Users\wey38795\Documents\mantid_builds\Master\raw\mantid\scripts\Interface\reduction_gui\widgets\sans\hfir_sample_data.py", line 64, in initialize_content
    if not self._has_instrumentview:
AttributeError: 'DirectBeam' object has no attribute '_has_instrumentview'
```

**Comment by RichardWaiteSTFC** [comment]
> Powder Diffraction Reduction and DGS Reduction are fine, but ORNL SANS still has HFIR which results in this error if chosen:
 ```
Thanks, I had this error once, but couldn't recreate it when I restarted workbench and started testing the interfaces again. Does it consistently do this for you?

**Comment by Harrietbrown** [comment]
it happens every time it is oppened unless you change the facility through another interface

**Comment by Harrietbrown** [comment]
The error with HFIR seems unrelated ot this specific issue, I have made a new issue for it #28088.
Add a release note for this and it will be ok.

**Review by Harrietbrown — DISMISSED** [review]
Good Job. The unwanted options are no longer in the dropdown menus of the GUI's and the other error from HFIR has had an issue made for it. It should be good to merge once Jenkins is finished with it.

File: scripts/Interface/reduction_application.py [hunk 1]
```diff
@@ -283,14 +284,16 @@ def _change_instrument(self):
...
+
         class InstrDialog(QDialog):
             def __init__(self, instrument_list=None):
                 QDialog.__init__(self)
                 self.ui = load_ui(__file__, 'ui/instrument_dialog.ui', baseinstance=self)
                 self.instrument_list = instrument_list
                 self.instr_combo.clear()
                 self.facility_combo.clear()
-                instruments = sorted(INSTRUMENT_DICT.keys())
+                instruments = sorted([fac for fac in INSTRUMENT_DICT.keys() if
```
Comment (martyngigg) [inline scripts/Interface/reduction_application.py]: Can I be picky and ask us to rename the `instruments` variable to `facilities`? I was trying to understand the list comprehension assuming it was supposed to be giving a list of instruments and it didn't make sense until I realised it wasn't.

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by martyngigg — DISMISSED** [review]

**Comment by martyngigg** [comment]
Just a [flake 8](https://builds.mantidproject.org/job/pull_requests-flake8/24972/flake8/new/source.ab4e43f6-8ede-470d-8b3b-0f4aad820a13/#296) issue to deal with here now.

**Review by martyngigg — APPROVED** [review]

**Pull request merged**